[DOC] Update TraceQL query editor doc

knylander-grafana opened the pull request:
**What this PR does**:
Clarifies the TraceQL query editor based upon input from @cyrille-leclerc (see #2607). 

**Which issue(s) this PR fixes**:
Fixes #2607 (https://github.com/grafana/tempo/issues/2607)

**Checklist**
- [ ] Tests updated
- [X] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

cyrille-leclerc commented on docs/sources/tempo/traceql/query-editor.md:
```diff
@@ -11,20 +11,63 @@ keywords:
+To query a particular trace:
+
+1. From the menu, choose **Explore** and select the **TraceQL** tab.
+1. Enter the trace’s ID into the query field.
```
Can we provide an example, is it between `{...}` brackets or without?

knylander-grafana replied:
I confirmed that adding a trace ID directly on the search field on the TraceQL tab and pressing Run query returns data for the trace ID. I have a screenshot I can add to this section, if you think that would be helpful.

cyrille-leclerc commented on docs/sources/tempo/traceql/query-editor.md:
```diff
@@ -11,20 +11,63 @@ keywords:
 ![Query editor showing request for http.method](/static/img/docs/tempo/query-editor-http-method.png)
 
-Using the query editor, you can use the editor’s autocomplete suggestions to write queries. The editor detects span sets to provide relevant autocomplete options. It uses regular expressions (regex) to detect where it is inside a spanset and provide attribute names, scopes, intrinsic names, logic operators, or attribute values from Tempo's API, depending on what is expected for the current situat ...
+### Query by TraceID
```
Can we add "Query by SpanID"? And also by time range? Maybe it's the selected time interval for the time range selection

knylander-grafana replied:
working on this today

knylander-grafana replied:
Okay, so query by Span ID does not work in the same way that Query by TraceID does. I'll figure out another medhot and write that up.

knylander-grafana replied:
I added an optional step to adjust the time range and linked to the documentation for the time picker.

knylander-grafana replied:
@cyrille-leclerc What do you think of the updates?

cyrille-leclerc commented on docs/sources/tempo/traceql/query-editor.md:
```diff
@@ -11,20 +11,66 @@ keywords:
 ---
 
-# TraceQL query editor
+# Create TraceQL queries in Grafana
```
```suggestion
# Write TraceQL queries in Grafana
```

"Create" sounds to me like something intended to last when here it's about exploring data, something not necessarily intended to be persisted.

knylander-grafana replied:
Thank you for the clarification.

cyrille-leclerc commented on docs/sources/tempo/traceql/query-editor.md:
```diff
@@ -11,20 +11,66 @@ keywords:
+# Create TraceQL queries in Grafana
 
-You can use the TraceQL viewer and query editor in the Tempo data source to build queries and drill-down into result sets. The editor is available in Grafana’s Explore interface.
+You can use compose TraceQL queries in Grafana and Grafana Cloud using the Search in Explore and the TraceQL viewer and query editor in the Tempo data source.
```
This sentence sounds to me like there is a "Tempo Datasource" screen that I can navigate to and write my TraceQL queries but I don't think that exists. It also suggest there is a view called "query editor", is it a reference to the trace query visual editor of the "Query Type" tab "Search"?

My understanding is that I have to
* Navigate to Explore
* Select a Tempo data source
* Navigate to the "TraceQL" tab. An alternative being to use the TraceQL visual builder of the "Search" tab.

knylander-grafana replied:
Thank you for the clarifications. I've reworded the content using your instructions. 

Query editor was the term used when we only had one location (TraceQL tab) to create TraceQL queries. I'll recheck the phrasing in this article ot make sure that the wording is correct. Thank you for the question.

cyrille-leclerc commented on docs/sources/tempo/traceql/query-editor.md:
```diff
@@ -11,20 +11,66 @@ keywords:
+The query editor and search are available in Grafana Cloud without enabling a feature flag.
 {{% /admonition %}}
 
+## Create TraceQL queries using the query editor
```
```suggestion
## Write TraceQL queries using the query editor
```

Additional question: what is the "Query Editor"? Is it the visual TraceQL builder of the "Query type" tab "Search"?

knylander-grafana replied:
This was a good point. We started off with only the TraceQL tab, so that was the TraceQL query editor. The Search tab was added on later, with the doc added to the same page. I've tried to clarify in this PR the differences. The query builder name is an excellent term to use to help clarify.

cyrille-leclerc commented on docs/sources/tempo/traceql/query-editor.md:
```diff
@@ -11,20 +11,66 @@ keywords:
+1. Sign into Grafana or Grafana Cloud.
+1. Select your Tempo data source.
+1. From the menu, choose **Explore** and select the **TraceQL** tab.
+1. Start your query on the text line by entering `{{`. For help with TraceQL syntax, refer to the [Construct a TraceQL query documentation]({{< relref "./_index.md" >}}).
```
```suggestion
1. Start your query on the text line by entering `{`. For help with TraceQL syntax, refer to the [Construct a TraceQL query documentation]({{< relref "./_index.md" >}}).
```

Isn't it one single bracket? `{`

knylander-grafana replied:
Good catch. I verified that it is a single `{` and not two. Updated.

cyrille-leclerc commented on docs/sources/tempo/traceql/query-editor.md:
```diff
@@ -11,20 +11,66 @@ keywords:
+
+To query a particular trace:
+
+1. From the menu, choose **Explore** and select the **TraceQL** tab.
```
```suggestion
1. From the menu, choose **Explore**, select the desired Tempo datasource and navigate to the **TraceQL** tab.
```

knylander-grafana replied:
Done. I'm editing the file directly to make changes.

cyrille-leclerc commented on docs/sources/tempo/traceql/query-editor.md:
```diff
@@ -11,20 +11,66 @@ keywords:
+
+1. From the menu, choose **Explore** and select the **TraceQL** tab.
+1. Enter the trace ID into the query field. For example: `1f187d8363b5a9b30cedd8e0ce9ccb43`
+1. Select **Run query**.
```
```suggestion
1. Click on **Run query** or use the keyboard shortcut shift+enter.
```

knylander-grafana replied:
I've updated the text. I tend to use Select instead of Click in case someone is using a touchscreen device like a tablet.

knylander-grafana commented on docs/sources/tempo/traceql/query-editor.md:
```diff
@@ -1,7 +1,7 @@
-title: TraceQL query editor
-menuTitle: TraceQL query editor
-description: Learn how to use the TraceQL query editor
+title: Create TraceQL queries in Grafana
+menuTitle: Create TraceQL queries in Grafana
```
```suggestion
title: Write TraceQL queries in Grafana
menuTitle: Write TraceQL queries in Grafana
```

knylander-grafana commented:
@cyrille-leclerc I think we're good to merge. What do you think?

cyrille-leclerc commented on docs/sources/tempo/traceql/query-editor.md:
```diff
@@ -11,20 +11,66 @@ keywords:
+# Write TraceQL queries in Grafana
 
-You can use the TraceQL viewer and query editor in the Tempo data source to build queries and drill-down into result sets. The editor is available in Grafana’s Explore interface.
+You can use compose TraceQL queries in Grafana and Grafana Cloud using **Explore** and a Tempo data source. You can use either the **Query type** > **Search** or the **TraceQL** tab.
```
I don't understand what _"You can use compose TraceQL queries in Grafana and Grafana Cloud"_ means. Should we remove "use"?

knylander-grafana replied:
It should be, You can compose. Thank you for the catch! 
```suggestion
You can compose TraceQL queries in Grafana and Grafana Cloud using **Explore** and a Tempo data source. You can use either the **Query type** > **Search** or the **TraceQL** tab.
```

cyrille-leclerc commented on docs/sources/tempo/traceql/query-editor.md:
```diff
@@ -11,20 +11,66 @@ keywords:
-Using the query editor, you can use the editor’s autocomplete suggestions to write queries. The editor detects span sets to provide relevant autocomplete options. It uses regular expressions (regex) to detect where it is inside a spanset and provide attribute names, scopes, intrinsic names, logic operators, or attribute values from Tempo's API, depending on what is expected for the current situat ...
+### Query by TraceID
+
+To query a particular trace:
```
```suggestion
To query a particular trace by its trace ID:
```

cyrille-leclerc commented:
@knylander-grafana I'm sorry, I propose two additional changes 👼

knylander-grafana commented:
@cyrille-leclerc I appreciate your thoroughness! Maybe this time will be the charm?

knylander-grafana commented:
With Tempo 2.2 RC being cut, we'll need to backport these changes to have them in the next release.

cyrille-leclerc commented on docs/sources/tempo/traceql/query-editor.md:
```diff
@@ -11,20 +11,66 @@ keywords:
-To use the TraceQL query editor in Grafana, you need to enable the `traceqlEditor` feature flag. This feature is available starting in Grafana 9.3.2. The query editory is available automatically in Grafana Cloud.
+To use the TraceQL query editor in Grafana 9.3.2 and newer, you need to enable the `traceqlEditor` feature flag.
+
+To enable the Trace Search in Grafana 10 and newer, you need to enable the `traceqlSearch` feature flag.
```
Comments that should not prevent from merging this PR:

1. I don't know where to edit Feature flags
2. Are we sue a feature flag must be changed? Are we sure TraceQL tab is not enabled by default in Grafana 10? I can see the TraceQL tab on my Grafana 10 on Grafana Cloud even though I don't remember tweaking my config

knylander-grafana replied:
I can add a link to the Feature Flag page easily enough. The feature flag isn't necessary in Grafana Cloud. i think both of these features are default in Cloud. But I will verify.

cyrille-leclerc replied:
Thank you @knylander-grafana for your pugnacity.
I think it would be great, in this doc update or a subsequent one, to be clear about the need or not of a feature flag on Grafana Cloud, especially for Grafana 10

joe-elliott replied:
I agree with @cyrille-leclerc I don't think that `traceqlSearch` is required in Grafana 10.

@grafana/observability-traces-and-profiling do we have a list of what feature flags are required in which Grafana versions for the traceql search and the traceql builder replacement of old search?

knylander-grafana replied:
@adrapereira It's still confusing on this page that we have two panels on the same page that deal with TraceQL search. The Query type > Search is the TraceQL query builder while the TraceQL tab is the query editor. Using these terms is confusing -- especially if our own people are getting confused by the terminology. 

How do you refer to the editor and the builder?

knylander-grafana replied:
Here's what I did: 
![image](https://github.com/grafana/tempo/assets/104772500/e48c089f-92f2-4f62-b501-eb6945a4010c)

cyrille-leclerc reviewed (approved):
Let's merge and follow up in subsequent PRs.
In addition to

1. I don't know where to edit Feature flags
2. Are we sue a feature flag must be changed? Are we sure TraceQL tab is not enabled by default in Grafana 10? I can see the TraceQL tab on my Grafana 10 on Grafana Cloud even though I don't remember tweaking my config

I think we have to refresh the screenshots as the tabs have changed now it is:
<img width="570" alt="image" src="https://github.com/grafana/tempo/assets/459691/3dc33411-5a98-4651-82dc-adedd16895f4">

knylander-grafana commented on docs/sources/tempo/traceql/query-editor.md:
```diff
@@ -11,20 +11,66 @@ keywords:
 
 {{% admonition type="note" %}}
-To use the TraceQL query editor in Grafana, you need to enable the `traceqlEditor` feature flag. This feature is available starting in Grafana 9.3.2. The query editory is available automatically in Grafana Cloud.
+To use the TraceQL query editor in Grafana 9.3.2 and newer, you need to enable the `traceqlEditor` feature flag.
```
```suggestion
To use the TraceQL query editor in Grafana 9.3.2 and newer, you need to enable the `traceqlEditor` feature toggle ([refer to instructions for enabling feature toggles](/docs/grafana/latest/setup-grafana/configure-grafana/feature-toggles/).)
```

joe-elliott commented on docs/sources/tempo/traceql/query-editor.md:
```diff
@@ -11,20 +11,66 @@ keywords:
 
 {{% admonition type="note" %}}
-To use the TraceQL query editor in Grafana, you need to enable the `traceqlEditor` feature flag. This feature is available starting in Grafana 9.3.2. The query editory is available automatically in Grafana Cloud.
+To use the TraceQL query editor in Grafana 9.3.2 and newer, you need to enable the `traceqlEditor` feature toggle ([refer to instructions for enabling feature toggles](/docs/grafana/latest/setup-grafana/configure-grafana/feature-toggles/).)
```
It's not clear to me what the "TraceQL query editor" is. I think this is the search replacement with the TraceQL builder. Does this have an official name?

knylander-grafana replied:
Oh! Then I need to remove the feature toggle information and clarifyi the text.

knylander-grafana replied:
Would it be better to say the TraceQL tab instead of the query editor since we now also have a query builder in Query type > Search?

joe-elliott reviewed (commented):
I'm struggling with the mix of cloud and grafana docs. Can we write this to just talk about how to use Grafana to write TraceQL queries?

I don't think we ever need to call out cloud specific anything. I know the goal is to share this doc between cloud and OSS but I think the message is a bit hard to follow with all the cloud vs. OSS caveats.

knylander-grafana commented:
> Let's merge and follow up in subsequent PRs. In addition to
> 
> 1. I don't know where to edit Feature flags
> 2. Are we sue a feature flag must be changed? Are we sure TraceQL tab is not enabled by default in Grafana 10? I can see the TraceQL tab on my Grafana 10 on Grafana Cloud even though I don't remember tweaking my config
> 
> I think we have to refresh the screenshots as the tabs have changed now it is: <img alt="image" width="570" src="https://user-images.githubusercontent.com/459691/255528111-3dc33411-5a98-4651-82dc-adedd16895f4.png">

I've updated the screenshots, so the tabs should be okay.

knylander-grafana commented:
@joe-elliott 
> I'm struggling with the mix of cloud and grafana docs. Can we write this to just talk about how to use Grafana to write TraceQL queries?
> 
> I don't think we ever need to call out cloud specific anything. I know the goal is to share this doc between cloud and OSS but I think the message is a bit hard to follow with all the cloud vs. OSS caveats.

Considering all of the conversation and comments about Cloud vs OSS feature toggles, I thought the same thing. All of that discussion was for a three-line Note -- and we had some excellent clarifications. 

I built the docs locally to check and realized might not be as bad as it seems. When I build the page locally, there are only 3 occurrences of the word Cloud in the page. 

I've updated the content to clarify the UI, updated all of the screenshots, and hopefully that will fix some of the issues.

knylander-grafana commented:
Here's what the current page looks like when built locally: 
![image](https://github.com/grafana/tempo/assets/104772500/3b19b008-38e7-47fe-a238-4f268adeb874)
